lib/opFns.js

```diff
@@ -157,6 +157,45 @@ module.exports = {
 
     return new BN(word.shrn((31 - pos.toNumber()) * 8).andln(0xff))
   },
+  SHL: function (a, b, runState) {
+    if (!runState._common.gteHardfork('constantinople')) {
+      trap(ERROR.INVALID_OPCODE)
+    }
+    if (a.gten(256)) {
+      return new BN(0)
+    }
+    return b.shln(a.toNumber()).iand(utils.MAX_INTEGER)
+  },
+  SHR: function (a, b, runState) {
+    if (!runState._common.gteHardfork('constantinople')) {
+      trap(ERROR.INVALID_OPCODE)
+    }
+    if (a.gten(256)) {
+      return new BN(0)
+    }
+    return b.shrn(a.toNumber())
+  },
+  SAR: function (a, b, runState) {
+    if (!runState._common.gteHardfork('constantinople')) {
+      trap(ERROR.INVALID_OPCODE)
+    }
+    const isSigned = b.testn(255)
+    if (a.gten(256)) {
+      if (isSigned) {
+        return new BN(utils.MAX_INTEGER)
+      } else {
+        return new BN(0)
+      }
+    }
+    const c = b.shrn(a.toNumber())
+    if (isSigned) {
+      const shiftedOutWidth = 255 - a.toNumber()
+      const mask = utils.MAX_INTEGER.shrn(shiftedOutWidth).shln(shiftedOutWidth)
+      return c.ior(mask)
+    } else {
+      return c
+    }
+  },
```

Comment:
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-145.md#0x1d-sar-arithmetic-shift-right

I also think I can roughly follow what is going on here and think all EIP conditions are covered and can also place all - this to me appearing a bit esoteric - mask stuff to the places in the EIP. Would have to do an extra hour of bit operation reading though for a better understanding.

Since tests are passing I will approve.